nixos/networkd: add `dhcpServerStaticLeaseConfig` option

veehaitch opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Add `systemd.network.networks.*.dhcpServerStaticLeaseConfig` to allow
for configuring static DHCP leases through the [`[DHCPServerStaticLease]`](https://www.freedesktop.org/software/systemd/man/systemd.network.html#%5BDHCPServerStaticLease%5D%20Section%20Options)
section. See systemd.network(5) of systemd 249 for details.

~Also extends the NixOS test `systemd-networkd-dhcpserver` to assign
static leases.~
Also adds the NixOS test `systemd-networkd-dhcpserver-static-lease` to
test the assignment of static leases.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

veehaitch commented:
@ofborg test systemd-networkd-dhcpserver

veehaitch commented:
@ofborg test systemd-networkd-dhcpserver

veehaitch commented:
@ofborg test systemd-networkd-dhcpserver

veehaitch commented:
@ofborg test systemd-networkd-dhcpserver

veehaitch commented:
I'm a bit puzzled why the NixOS test `systemd-networkd-dhcpserver` fails. The client receives an IPv4 address but it's not the one statically assigned.

The systemd configurations look good to me:

```ini
# router: /etc/systemd/network/01-eth1.network
[Match]                                                                                                                                                                                                     Name=eth1

[Network]                                                                                                                                                                                                                           Address=10.0.0.1/24                                                                                                                                                                                         DHCPServer=true                                                                                                                                                                                                                                                                                                                                                                                                                                                         [DHCPServerStaticLease]                                                                                                                                                                                                             Address=10.0.0.10                                                                                                                                                                                                                   MACAddress=02:de:ad:be:ef:01
```

```ini
# client: /etc/systemd/network/40-eth1.network
[Match]
Name=eth1

[Link]
MACAddress=02:de:ad:be:ef:01

[Network]
DHCP=yes
IPv6PrivacyExtensions=kernel
Address=192.168.1.1/24
```

Marking this as ready for review to collect further opinions why this fails.

tomfitzhenry commented on nixos/tests/systemd-networkd-dhcpserver.nix:
```diff
@@ -28,10 +28,12 @@ import ./make-test-python.nix ({pkgs, ...}: {
               DHCPServer = true;
               Address = "10.0.0.1/24";
             };
-            dhcpServerConfig = {
-              PoolOffset = 100;
-              PoolSize = 1;
-            };
```
This would remove the testing of dynamic leases.

It might be worth keeping this dynamic lease test, and creating a separate test for static leases?

veehaitch replied:
Agreed, that makes sense. At the moment, however, it doesn't work at all 😐.

veehaitch replied:
I moved the test to a new file.

tomfitzhenry commented:
Re the test failure, it'd be interesting to see the bytes on the wire between `client` and `router`, to see which MAC address is sent. I don't know how to do that. Can a delay be added to startup, during which time you could SSH in? Or perhaps the VM could run tcpdump at early boot?

veehaitch commented:
I intercepted all the UDP port 67 and 68 traffic on `router`. Unfortunately, I still don't see why it wouldn't work.

```shell
# tcpdump -i eth1 "udp port 67 or udp port 68"
16:23:06.173849 IP (tos 0xc0, ttl 64, id 0, offset 0, flags [none], proto UDP (17), length 316)
    0.0.0.0.68 > 255.255.255.255.67: BOOTP/DHCP, Request from 02:de:ad:be:ef:01, length 288, xid 0x294e07b2, secs 1, Flags [none]
          Client-Ethernet-Address 02:de:ad:be:ef:01
          Vendor-rfc1048 Extensions
            Magic Cookie 0x63825363
            DHCP-Message Option 53, length 1: Discover
            Client-ID Option 61, length 19: hardware-type 255, b7:b0:95:87:00:02:00:00:ab:11:44:3b:62:af:cf:49:d4:de
            Parameter-Request Option 55, length 9:
              Subnet-Mask, Default-Gateway, Domain-Name-Server, Hostname
              Domain-Name, Static-Route, NTP, Option 120
              Classless-Static-Route
            MSZ Option 57, length 2: 576
            Hostname Option 12, length 6: "client"
16:23:06.188218 IP (tos 0xc0, ttl 64, id 0, offset 0, flags [none], proto UDP (17), length 296)
    10.0.0.1.67 > 10.0.0.190.68: BOOTP/DHCP, Reply, length 268, xid 0x294e07b2, Flags [none]
          Your-IP 10.0.0.190
          Client-Ethernet-Address 02:de:ad:be:ef:01
          Vendor-rfc1048 Extensions
            Magic Cookie 0x63825363
            DHCP-Message Option 53, length 1: Offer
            Lease-Time Option 51, length 4: 3600
            Subnet-Mask Option 1, length 4: 255.255.255.0
            Default-Gateway Option 3, length 4: 10.0.0.1
            Server-ID Option 54, length 4: 10.0.0.1
16:23:06.242605 IP (tos 0xc0, ttl 64, id 0, offset 0, flags [none], proto UDP (17), length 328)
    0.0.0.0.68 > 255.255.255.255.67: BOOTP/DHCP, Request from 02:de:ad:be:ef:01, length 300, xid 0x294e07b2, secs 1, Flags [none]
          Client-Ethernet-Address 02:de:ad:be:ef:01
          Vendor-rfc1048 Extensions
            Magic Cookie 0x63825363
            DHCP-Message Option 53, length 1: Request
            Client-ID Option 61, length 19: hardware-type 255, b7:b0:95:87:00:02:00:00:ab:11:44:3b:62:af:cf:49:d4:de
            Parameter-Request Option 55, length 9:
              Subnet-Mask, Default-Gateway, Domain-Name-Server, Hostname
              Domain-Name, Static-Route, NTP, Option 120
              Classless-Static-Route
            MSZ Option 57, length 2: 576
            Server-ID Option 54, length 4: 10.0.0.1
            Requested-IP Option 50, length 4: 10.0.0.190
            Hostname Option 12, length 6: "client"
16:23:06.245152 IP (tos 0xc0, ttl 64, id 0, offset 0, flags [none], proto UDP (17), length 301)
    10.0.0.1.67 > 10.0.0.190.68: BOOTP/DHCP, Reply, length 273, xid 0x294e07b2, Flags [none]
          Your-IP 10.0.0.190
          Client-Ethernet-Address 02:de:ad:be:ef:01
          Vendor-rfc1048 Extensions
            Magic Cookie 0x63825363
            DHCP-Message Option 53, length 1: ACK
            Lease-Time Option 51, length 4: 3600
            Subnet-Mask Option 1, length 4: 255.255.255.0
            Default-Gateway Option 3, length 4: 10.0.0.1
            TZ-Name Option 101, length 3: "UTC"
            Server-ID Option 54, length 4: 10.0.0.1
```

The `systemd-networkd` log output, although using a debug level, is not particularly helpful:

```
client # [   87.441352] systemd-networkd[486]: eth1: LLDP: Started LLDP client
client # [   87.450405] systemd-networkd[486]: eth1: Started LLDP.
client # [   87.476871] systemd-networkd[486]: eth1: Gained carrier
client # [   87.496019] systemd-networkd[486]: eth1: Acquiring DHCPv4 lease
client # [   87.517864] systemd-networkd[486]: eth1: DHCPv4 client: STARTED on ifindex 3
client # [   87.543969] systemd-networkd[486]: eth1: Requesting address: 192.168.1.1/24 (valid forever, preferred forever), flags: n/a
router # [  193.162496] systemd-networkd[486]: eth1: DHCPv4 server: DISCOVER (0x294e07b2)
router # [  193.169192] systemd-networkd[486]: eth1: DHCPv4 server: OFFER (0x294e07b2)
router # [  193.223725] systemd-networkd[486]: eth1: DHCPv4 server: REQUEST (selecting) (0x294e07b2)
router # [  193.229719] systemd-networkd[486]: eth1: DHCPv4 server: ACK (0x294e07b2)
```

tomfitzhenry commented:
Interesting. It looks like systemd is not behaving according to the DHCPServerStaticLease config: the DHCPDISCOVER has the expected MAC, but the DHCPOFFER does not have the expected IP.

A few hypotheses:
1. The version of systemd does not support `DHCPServerStaticLease`
2. There is a bug in systemd.

I vaguely recall that when systemd doesn't recognize a directive it sometimes (?) logs this without the unit field, i.e. `journalctl -u systemd-networkd.service` would not find it. `journalctl | -i dhcp` might be worth a look?

veehaitch commented:
> I vaguely recall that when systemd doesn't recognize a directive it sometimes (?) logs this without the unit field, i.e. `journalctl -u systemd-networkd.service` would not find it. `journalctl | -i dhcp` might be worth a look?

Unfortunately, there's nothing in the logs that would indicate an invalid configuration file. On the other hand, there's also no message which says setting up the `DHCPServerStaticLease` worked.

veehaitch commented:
Related upstream issue (also reported using NixOS): https://github.com/systemd/systemd/issues/21368

veehaitch committed:
nixos/networkd: add `dhcpServerStaticLeaseConfig` option

Add `systemd.network.networks.*.dhcpServerStaticLeaseConfig` to allow
for configuring static DHCP leases through the `[DHCPServerStaticLease]`
section. See systemd.network(5) of systemd 249 for details.

Also adds the NixOS test `systemd-networkd-dhcpserver-static-lease` to
test the assignment of static leases.

veehaitch commented:
@ofborg test systemd-networkd-dhcpserver-static-leases

veehaitch commented:
Thanks to a comment in the referenced upstream issue, this works now. Apparently, the client has to set `ClientIdentifier=` to `mac`; the default is `duid`:

https://github.com/NixOS/nixpkgs/blob/ce81231420ec3499ac4e8f31076cc9425dbf90f3/nixos/tests/systemd-networkd-dhcpserver-static-leases.nix#L57

I also added a comment to the `dhcpServerStaticLeaseConfig`:

https://github.com/NixOS/nixpkgs/blob/ce81231420ec3499ac4e8f31076cc9425dbf90f3/nixos/modules/system/boot/networkd.nix#L1186-L1187

Additionally, I created a dedicated test `systemd-networkd-dhcpserver-static-leases` instead of modifying the `systemd-networkd-dhcpserver` test.

tomfitzhenry commented:
> Apparently, the client has to set ClientIdentifier= to mac; the default is duid

Ah. https://datatracker.ietf.org/doc/html/rfc4361#section-6.3 suggests "DHCPv4 servers MAY use administrator-supplied values for chaddr and htype to identify the client in the case where the administrator is assigning a fixed IP address to the client, even if the client sends a client identifier option."

It seems like systemd-networkd-dhcpserver chooses not to implement that `MAY`, which then doesn't manage to identify the systemd DHCP client which uses DUID as the default client identifier (option 61).

Though that's a bit inconvenient, I think the intent here is that folks move to DUID-based identifiers.

Regardless, this PR LGTM. Good to see a test being added for it, and finding this issue!
